reconcile/utils/secret_reader.py

```diff
@@ -18,7 +21,10 @@ class SecretNotFound(Exception):
 class SecretReader:
     """Read secrets from either Vault or a config file."""
 
-    def __init__(self, settings: Optional[Mapping] = None) -> None:
+    def __init__(
+        self,
+        settings: Optional[Union[Mapping, AppInterfaceSettingsV1]] = None,
```

Comment:
The MR for best practices when using data query classes is still[ in progress](https://gitlab.cee.redhat.com/service/dev-guidelines/-/merge_requests/134/diffs) and I've done a poor job of contributing to it. With that being said, anything it `/utils/` feels like a `Inside shared functions` case from that doc. I understand the fragments are meant to minimize having many versions of gql data classes that all need to be supported by our libraries, but it's still coupled to both gql and the current implementation of how we generate data classes.

I'll contribute a better example for `Protocols` to our docs, but I wanted to get your thoughts on something like:

```python
# Open to better naming, many Protocol examples in the PEP use the "Supports" prefix when checking
# for the existence of a certain method, but this could also be HasVaultToggle or anything else.
# This protocol would be defined in the same module that uses it.
class SupportsVault(Protocol):
    vault: bool

# And then your __init__ changes to:
settings: Optional[Union[Mapping, SupportsVault]] = None,
```
The primary benefit being that you've now defined exactly what your code is looking for and the type definition would only ever need to change if your code changes. If some subtle variation of `AppInterfaceSettingsV1` is added (I know we'd prefer this doesn't happen, but it could), then there's no need to change this code as long as `vault` is defined and is a `bool`.

Reply:
Make total sense to me, and I see the pros and cons of both ways. Let's use the `Protocol` pattern, for now, which seems to be the better one :) TBH, I don't have a strong opinion on that, and I won't die on that hill ;)